pkg/controller/scylladbmanagerclusterregistration/controller.go

```diff
@@ -0,0 +1,356 @@
+	}
+}
+
+func (smcrc *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+
+	klog.InfoS("Starting controller", "controller", ControllerName)
+
+	var wg sync.WaitGroup
+	defer func() {
+		klog.InfoS("Shutting down controller", "controller", ControllerName)
+		smcrc.queue.ShutDown()
+		wg.Wait()
+		klog.InfoS("Shut down controller", "controller", ControllerName)
+	}()
+
+	if !cache.WaitForNamedCacheSync(ControllerName, ctx.Done(), smcrc.cachesToSync...) {
+		return
+	}
+
+	for range workers {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			wait.UntilWithContext(ctx, smcrc.runWorker, time.Second)
+		}()
+	}
+
+	<-ctx.Done()
+}
+
+func (smcrc *Controller) addScyllaDBManagerClusterRegistration(obj interface{}) {
+	smcrc.handlers.HandleAdd(
+		obj.(*scyllav1alpha1.ScyllaDBManagerClusterRegistration),
+		smcrc.handlers.Enqueue,
+	)
+}
+
+func (smcrc *Controller) updateScyllaDBManagerClusterRegistration(old, cur interface{}) {
+	smcrc.handlers.HandleUpdate(
+		old.(*scyllav1alpha1.ScyllaDBManagerClusterRegistration),
+		cur.(*scyllav1alpha1.ScyllaDBManagerClusterRegistration),
+		smcrc.handlers.Enqueue,
+		smcrc.deleteScyllaDBManagerClusterRegistration,
+	)
+}
+
+func (smcrc *Controller) deleteScyllaDBManagerClusterRegistration(obj interface{}) {
+	smcrc.handlers.HandleDelete(
+		obj,
+		smcrc.handlers.Enqueue,
+	)
+}
+
+func (smcrc *Controller) addScyllaDBDatacenter(obj interface{}) {
+	smcrc.handlers.HandleAdd(
+		obj.(*scyllav1alpha1.ScyllaDBDatacenter),
+		smcrc.enqueueThroughScyllaDBDatacenter,
+	)
+}
+
+func (smcrc *Controller) updateScyllaDBDatacenter(old, cur interface{}) {
+	smcrc.handlers.HandleUpdate(
+		old.(*scyllav1alpha1.ScyllaDBDatacenter),
+		cur.(*scyllav1alpha1.ScyllaDBDatacenter),
+		smcrc.enqueueThroughScyllaDBDatacenter,
+		smcrc.deleteScyllaDBDatacenter,
+	)
+}
+
+func (smcrc *Controller) deleteScyllaDBDatacenter(obj interface{}) {
+	smcrc.handlers.HandleDelete(
+		obj,
+		smcrc.enqueueThroughScyllaDBDatacenter,
+	)
+}
+
+func (smcrc *Controller) addSecret(obj interface{}) {
+	smcrc.handlers.HandleAdd(
+		obj.(*corev1.Secret),
+		smcrc.enqueueThroughOwner,
+	)
+}
+
+func (smcrc *Controller) updateSecret(old, cur interface{}) {
+	smcrc.handlers.HandleUpdate(
+		old.(*corev1.Secret),
+		cur.(*corev1.Secret),
+		smcrc.enqueueThroughOwner,
+		smcrc.deleteSecret,
+	)
+}
+
+func (smcrc *Controller) deleteSecret(obj interface{}) {
+	smcrc.handlers.HandleDelete(
+		obj.(*corev1.Secret),
+		smcrc.enqueueThroughOwner,
+	)
+}
+
+func (smcrc *Controller) addNamespace(obj interface{}) {
+	smcrc.handlers.HandleAdd(
+		obj.(*corev1.Namespace),
+		smcrc.handlers.EnqueueAllWithUntypedFilterFunc(hasGlobalScyllaDBManagerNamespaceName),
+	)
+}
+
+func (smcrc *Controller) updateNamespace(old, cur interface{}) {
+	smcrc.handlers.HandleUpdate(
+		old.(*corev1.Namespace),
+		cur.(*corev1.Namespace),
+		smcrc.handlers.EnqueueAllWithUntypedFilterFunc(hasGlobalScyllaDBManagerNamespaceName),
+		smcrc.deleteNamespace,
```

Comment:
EnqueueAll enqueues all ScyllaDBManagerClusterRegistration objects from untypedObj namespace. But this is Namespace so it has empty .Namespace field, which value turns out to be equal to special corev1.NamespaceAll, meaninig it will enqueue all ScyllaDBManagerClusterRegistration from all namespaces. Is this expected behavior?

Reply:
Yes, e.g. deletion of `scylla-manager` namespace should unblock finalization of all existing SMCRs having the global-manager label. Although I can write a dedicated method if you think the route to this behaviour is hacky

Reply:
it's very much hacky, lets have dedicated method expressing what is desired behavior

Reply:
done